permissions.ts

```diff
@@ -0,0 +1,68 @@
+import type { JwtPayload } from "jsonwebtoken";
+import type { KeycloakTokenParsed } from "keycloak-js";
+import { Request as JwtRequest } from "express-jwt";
+import { NextFunction, RequestHandler, Response } from "express";
+
+import { State } from "./globals.js";
+
+export type KeycloakJwtRequest = JwtRequest<JwtPayload & KeycloakTokenParsed>;
+
+export type ConstellationsRole = 'update-home-timeline' | 'update-global-tessellation' |
+                                 'manage-handles' | 'manage-features';
+
+export function amISuperuser(req: JwtRequest, state: State): boolean {
+  return req.auth !== undefined && req.auth.sub === state.config.superuserAccountId;
+}
+
+export function hasRole(req: KeycloakJwtRequest, role: ConstellationsRole): boolean {
+  return req.auth !== undefined && !!req.auth.realm_access?.roles.includes(role);
+}
+
+export function makeRequireRoleMiddleware(role: ConstellationsRole): RequestHandler {
+  return (req: KeycloakJwtRequest, res: Response, next: NextFunction) => {
+    if (!hasRole(req, role)) {
+      res.status(403).json({
+        error: true,
+        message: "Forbidden",
+      });
+    }
+    next();
```

Comment:
I'm not 100% sure about the semantics of how these handlers work, but I think this should be an `else { next() ; }`, no? Even if the `next()` ends up not doing anything in the Forbidden case, it feels safer to not invoke it at all ...